File: myfidpass/Helpers/CommerceFlyerStore.swift
//
//  CommerceFlyerStore.swift
//  myfidpass
//
//  Source de vérité unifiée (mémoire) pour l’état flyer cross-écrans.
//

import Foundation

final class CommerceFlyerStore {
    static let shared = CommerceFlyerStore()

    struct Snapshot: Equatable {
        var flyerRegistered: Bool
        var shareURL: String
        var bootstrapPreviewB64: String?
        var customBgDataURL: String?
        var revisionKey: String?
    }

    private var snapshotsBySlug: [String: Snapshot] = [:]

    private init() {}

    func snapshot(for slug: String) -> Snapshot? {
        let key = slug.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return nil }
        return snapshotsBySlug[key]
    }

    func hydrateFromDiskIfNeeded(slug: String) {
        let key = slug.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        guard snapshotsBySlug[key] == nil else { return }
        guard let cached = CommerceFlyerStateCache.load(slug: key) else { return }
        snapshotsBySlug[key] = Snapshot(
            flyerRegistered: cached.flyerRegistered,
            shareURL: cached.shareURL,
            bootstrapPreviewB64: cached.bootstrapPreviewB64,
            customBgDataURL: cached.customBgDataURL,
            revisionKey: cached.revisionKey
        )
    }

    func upsert(slug: String, snapshot: Snapshot) {
        let key = slug.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        snapshotsBySlug[key] = snapshot
        NotificationCenter.default.post(name: .myfidpassMerchantSetupProgressUpdated, object: nil)
    }

    /// Flyer personnalisé (cache, mémoire ou brouillon éditeur) — même critère que l’accueil / checklist.
    func isFlyerReady(for slug: String?) -> Bool {
        let key = slug?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !key.isEmpty else { return false }
        Self.shared.hydrateFromDiskIfNeeded(slug: key)
        if let snap = Self.shared.snapshot(for: key) {
            if snapLooksReady(snap) { return true }
        }
        if let draft = CommerceFlyerEditorDraftStore.load(slug: key) {
            if !draft.bootstrapB64.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return true }
            let draftBg = draft.meta.customBgDataURL?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !draftBg.isEmpty { return true }
        }
        guard let cached = CommerceFlyerStateCache.load(slug: key) else { return false }
        let hasBootstrap = !(cached.bootstrapPreviewB64 ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let hasCustomBg = !(cached.customBgDataURL ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return cached.flyerRegistered || hasBootstrap || hasCustomBg
    }

    private func snapLooksReady(_ snap: Snapshot) -> Bool {
        let hasBootstrap = !(snap.bootstrapPreviewB64 ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let hasCustomBg = !(snap.customBgDataURL ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return snap.flyerRegistered || hasBootstrap || hasCustomBg
    }

    func clear(slug: String) {
        let key = slug.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        snapshotsBySlug.removeValue(forKey: key)
    }

    func clearAll() {
        snapshotsBySlug.removeAll()
    }
}

File: android/app/src/main/java/fr/myfidpass/data/local/CommerceFlyerStore.kt
package fr.myfidpass.data.local

import android.content.Context
import fr.myfidpass.flyer.FlyerBootstrapPreviewPayloadBuilder

/** Mémoire cross-écrans — aligné iOS `CommerceFlyerStore`. */
object CommerceFlyerStore {

    data class Snapshot(
        val flyerRegistered: Boolean = false,
        val shareURL: String = "",
        val bootstrapPreviewB64: String? = null,
        val customBgDataURL: String? = null,
        val revisionKey: String? = null,
    )

    private val memory = mutableMapOf<String, Snapshot>()

    fun snapshot(slug: String): Snapshot? = memory[slug.trim()]

    fun hydrateFromDiskIfNeeded(context: Context, slug: String) {
        val key = slug.trim()
        if (key.isEmpty() || memory.containsKey(key)) return
        CommerceFlyerStateCache.load(context, key)?.let { cached ->
            memory[key] = Snapshot(
                flyerRegistered = cached.flyerRegistered,
                shareURL = cached.shareURL,
                bootstrapPreviewB64 = cached.bootstrapPreviewB64,
                customBgDataURL = cached.customBgDataURL,
                revisionKey = cached.revisionKey,
            )
        }
    }

    fun update(context: Context, slug: String, snapshot: Snapshot) {
        val key = slug.trim()
        if (key.isEmpty()) return
        memory[key] = snapshot
        CommerceFlyerStateCache.save(
            context = context,
            slug = key,
            flyerRegistered = snapshot.flyerRegistered,
            shareURL = snapshot.shareURL,
            bootstrapPreviewB64 = snapshot.bootstrapPreviewB64,
            customBgDataURL = snapshot.customBgDataURL,
            revisionKey = snapshot.revisionKey,
        )
    }

    fun isFlyerReady(context: Context, slug: String): Boolean {
        hydrateFromDiskIfNeeded(context, slug)
        val snap = snapshot(slug) ?: CommerceFlyerStateCache.load(context, slug)?.let { cached ->
            Snapshot(
                flyerRegistered = cached.flyerRegistered,
                shareURL = cached.shareURL,
                bootstrapPreviewB64 = cached.bootstrapPreviewB64,
                customBgDataURL = cached.customBgDataURL,
                revisionKey = cached.revisionKey,
            )
        }
        if (snap != null) {
            if (snap.flyerRegistered) return true
            if (!snap.bootstrapPreviewB64.isNullOrBlank()) return true
            if (!snap.customBgDataURL.isNullOrBlank()) return true
        }
        return CommerceFlyerEditorDraftStore.load(context, slug) != null
    }

    fun clearAll(context: Context) {
        memory.clear()
        CommerceFlyerStateCache.clearAll(context)
    }
}
